src/runtime/heap/id.c

```diff
@@ -110,10 +122,13 @@ static u64 id_alloc_from_range(id_heap i, id_range r, u64 pages, range subrange)
     u64 max_start = range_span(ri) > pages_rounded ?
         (range_span(ri) & ~(pages_rounded - 1)) - pages_rounded : 0;
     u64 start_bit = ri.start;
-    if ((i->flags & ID_HEAP_FLAG_RANDOMIZE) && max_start > 0)
+    if ((i->flags & ID_HEAP_FLAG_RANDOMIZE) && max_start > 0) {
         start_bit += random_u64() % max_start;
-    else if (point_in_range(ri, r->next_bit))
-        start_bit = r->next_bit;
+    } else {
+        u64 next_bit = get_next_bit(r, page_order);
+        if (point_in_range(ri, next_bit))
+            start_bit = MIN(next_bit, start_bit + max_start);
```

Comment:
Trying to understand why max_start is added to start_bit here... is the reasoning that, if next_bit extends past the acceptable range from which we can allocate, then we want to start at the highest possible bit to avoid first-fit-like overhead?

Reply:
I understand that max_start (margin) is relative. But I thought it was only necessary for the randomize computation, so I was wondering why it is being added here (i.e. instead of the assignment being ```MIN(next_bit, start_bit)```).

Reply:
MIN(next_bit, start_bit) would always equal start_bit, so we wouldn't be using next_bit at all.
On the other end, we cannot just assign start_bit = next_bit because if next_bit is not aligned to the alloc size (which can happen if next_bit is being borrowed from a smaller alloc size) we may not be able to alloc even if there is enough room (because we do first bitmap_alloc_within_range(start_bit, ri.end) and then bitmap_alloc_within_range(ri.start, start_bit)).